video/vdu_buffered.h

```diff
@@ -1060,4 +1072,203 @@ void VDUStreamProcessor::bufferCopyAndConsolidate(uint16_t bufferId, std::vector
+			auto bufferLength = block->size();
+			auto p_data = block->getBuffer();
+			orig_size += bufferLength;
+			debug_log(" from buffer %u [%08X] %u bytes\n\r", sourceBufferId, p_data, bufferLength);
+			debug_log(" %02hX %02hX %02hX %02hX\n\r",
+						p_data[0], p_data[1], p_data[2], p_data[3]);
+			cd.input_count += bufferLength;
+			while (bufferLength--) {
+				agon_compress_byte(&cd, *p_data++);
+			}
+		}
+		agon_finish_compression(&cd);
+
+		// make a single buffer with all of the temporary output data
+		auto bufferStream = make_shared_psram<BufferStream>(cd.output_count);
+		if (!bufferStream || !bufferStream->getBuffer()) {
+			// buffer couldn't be created
+			debug_log("bufferCompress: failed to create buffer %d\n\r", bufferId);
+			return;
+		}
+
+		auto p_hdr = (CompressionFileHeader*) p_temp;
+		p_hdr->orig_size = orig_size;
+
+		auto destination = bufferStream->getBuffer();
+		debug_log(" %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX\n\r",
+					p_temp[0], p_temp[1], p_temp[2], p_temp[3],
+					p_temp[4], p_temp[5], p_temp[6], p_temp[7],
+					p_temp[8], p_temp[9], p_temp[10], p_temp[11]);
+		memcpy(destination, p_temp, cd.output_count);
+		heap_caps_free(p_temp);
+		buffers[bufferId].push_back(bufferStream);
+
+		uint32_t pct = (cd.output_count * 100) / cd.input_count;
+		debug_log("Compressed %u input bytes to %u output bytes (%u%%) at %08X\n\r",
+				cd.input_count, cd.output_count, pct, destination);
+	} else {
+		debug_log("bufferCompress: cannot allocate temporary buffer of %d bytes\n\r", COMPRESSION_OUTPUT_CHUNK_SIZE);
+	}
+}
+
+// VDU 23, 0, &A0, bufferId; &1D, sourceBufferId; sourceBufferId; ...; 65535; : Decompress blocks from buffers
+// Decompress (blocks from) a list of buffers into a new buffer.
+// List is terminated with a bufferId of 65535 (-1).
+// Replaces the target buffer with the new one.
+// If target buffer is included in the source list it will be skipped.
+//
+void VDUStreamProcessor::bufferDecompress(uint16_t bufferId, uint16_t sourceBufferId) {
+	// Validate the compression header
+	if (buffers[sourceBufferId][0]->size() < sizeof(CompressionFileHeader)) {
+		debug_log("bufferDecompress: buffer too small for header\n\r");
+		return;
+	}
+	
+	auto p_hdr = (const CompressionFileHeader*) buffers[sourceBufferId][0]->getBuffer();
+	if (p_hdr->marker[0] != 'C' ||
+		p_hdr->marker[1] != 'm' ||
+		p_hdr->marker[2] != 'p' ||
+		p_hdr->type != COMPRESSION_TYPE_TURBO) {
+		debug_log("bufferDecompress: header is invalid\n\r");
+		return;
+	}
+	auto orig_size = p_hdr->orig_size;
+
+	buffers[bufferId].clear();
+	debug_log("Decompressing into buffer %u\n\r", bufferId);
+
+	// create a temporary output buffer, which may be expanded during decompression
+	uint8_t* p_temp = (uint8_t*) ps_malloc(COMPRESSION_OUTPUT_CHUNK_SIZE);
+	if (p_temp) {
+		// prepare for doing compression
+		DecompressionData dd;
+		agon_init_decompression(&dd, &p_temp, &local_write_decompressed_byte);
```

Comment:
just an idea here....

maybe this might be better as:
```
DecompressionData dd = agon_init_decompression();
```

if I've understood correctly `p_temp` essentially can be accessed as `dd->context` - `agon_init_decompression` could do it's own malloc thingy for it's output buffer, negating the need to create the `p_temp` block in this function.  also as suggested above the `local_write_decompressed_byte` function may be better off inside `compression.h`

Reply:
Agree with the last statement, as it's similar to the sister function.

The point behind "context" was to allow the compression/decompression operations to be performed for any reason desired by future code, not specifically related to buffers or memory allocation. And the use of local_write_compressed_byte and its sister function are by definition a part of the overall context of the operation(s). IOW, this use of compress/decompress for buffers is just 1 example or instance of using those operations, and I wanted to leave the "compression.h" functions open for custom contexts. That's why the function pointers and context value are "installed" at runtime, and not coded directly. Does that make sense? It's a "be generic" kind of thing.

Reply:
yeah - makes sense

I guess the direction I was heading with my comment was essentially "turn `agon_init_decompression` into a constructor for a `DecompressionData` class".  heading a little further in that direction and one can have the constructor have default arguments (and/or alternate constructors), allowing for a buffer to be passed into it, etc - giving that "be generic" thingy